File: Branch Sums/recursive.js
// O(n) time and space
class BinaryTree {
    constructor(value) {
      this.value = value;
      this.left = null;
      this.right = null;
    }
  }
  
  function branchSums(root) {
    // Write your code here.
    let branchSums=[]
    calculateBranchSums(branchSums, root, 0)
    return branchSums
  }
  
  const calculateBranchSums = (branchSums, node, currentSum) => {
    
    if (!node) return 
  
    currentSum += node.value 
    if (!node.right && !node.left){
      branchSums.push(currentSum)
    }
  
    calculateBranchSums(branchSums, node.left, currentSum)
    calculateBranchSums(branchSums, node.right, currentSum)
  }

File: Branch Sums/recursive.py
# O(n) time and space
class BinaryTree:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None


def branchSums(root):
    # Write your code here.
    branchSums=[]
    calculateBranchSums(branchSums, root,0)
    return branchSums

def calculateBranchSums(branchSums,currentNode, currentSum):
    if currentNode is None: return
    currentSum += currentNode.value
    if currentNode.left is None and  currentNode.right is None:
        branchSums.append(currentSum)
        return
    calculateBranchSums(branchSums,currentNode.left, currentSum)
    calculateBranchSums(branchSums,currentNode.right, currentSum)